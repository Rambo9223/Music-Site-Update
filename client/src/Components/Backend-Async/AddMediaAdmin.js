// add media async function for admin
const AddMediaAdmin = async(item,token) => {
    // item to add & authorisation 
    //post to server
    const response = await fetch("/admin/media/new", {
       method: "POST",
       headers: {
         "Content-Type": "application/json",
         token:token
       },
       body:JSON.stringify(item),
     });
     // return response.json regardless of outcome
     if (response.status !== 200) {
       return response.json();
     } else {
       return response.json();
     }
   };

   export default AddMediaAdmin