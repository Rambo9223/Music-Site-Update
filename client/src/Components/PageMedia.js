import { useState,useEffect } from "react"
import { Button,Card,ListGroup } from "react-bootstrap";
import FindMediaAdmin from "./Backend-Async/FindMediaAdmin";
import DeleteMedia from "./Backend-Async/DeleteMedia";
import EditPageMedia from "./EditPageMedia";
import Logo from "../Images/Logo.png";
import PopUp from "./PopUp";
import {Trash,Folder2Open,PlusCircle,QuestionCircle } from "react-bootstrap-icons";
import AddPageMedia from "./AddPageMedia";

// page displaying page media for admins to view
export default function PageMedia(props){

    let token = props.token;// auth token 
    let initialPopUp = {// popup
      message:"",
      type:"",
      response:""
    }

    const [media,setMedia] = useState([]);// array for media object 
    const [filter,setFilter] = useState(null);// search filter
    const [edit,setEdit] = useState(null);// contains item details for user edit 
    const [newMedia,setNewMedia] = useState(false);// opens add page media form
    const [popup,setPopUp] = useState(initialPopUp)// popup 
    const [bool,setBool] = useState(false);// bool to toggle popup 
    const ToggleBool = () => {setBool(!bool);
    if(bool===true){setPopUp(initialPopUp);setEdit(null)};setNewMedia(false)};
    
    // retrieve media object , depending on filter supplied 
    useEffect(()=>{
      if(filter!==null){
      FindMediaAdmin(filter,token).then((res)=>{
        if(res.ok===true){
          setMedia(res.results);// on success set results to setMedia hook 
        }else{
          // reset filter and bring up popup with error message 
          setBool(true);
          setPopUp({message:res.message,type:"Server Response",response:"text-danger"})
          setFilter(null);
        }
      })}
    },);

    // function to retrieve filter information from page
    function HandleFilter(){
      if(document.getElementById("filters").value===""){// no filter
        setFilter({query:"",
        filter:""});
      }else{
      setFilter({query:"media",// set to supplied filter 
      filter:document.getElementById("filters").value});
      }
    }

    // on edit open form and add details to fields
    function HandleEdit(item){
      setBool(true);
      setEdit(item);
    }

    return (
    <>
    <h3 className="faded-title">Page Media</h3>
    <div>
    <Button variant="success" style={{"marginTop":"0.5%"}} onClick={()=>{ToggleBool(); setNewMedia(true)}} >Add Media <PlusCircle/></Button>
    
    <h4 className="faded-title">Show Media:</h4>
    <select  className="form-select-md" id="filters" aria-label="Default select example">
    <option defaultValue={""}>Filter by:</option>
    <option value={""}>All</option>
    <option value="image">Images</option>
    <option value="video">Video</option>
    <option value="logo">Social Media</option>
</select>
<br/> 
<Button style={{marginTop:"0.5%"}} onClick={()=>{HandleFilter()}} type="button" className="btn btn-primary">
  Search <QuestionCircle/>{/*<Search/>*/}
</Button>
  </div>
  {media.length>0?<h4 className="faded-title">{media.length} result{(media.length>1)?<>s</>:null} found:</h4>:null}
  <div className="gallery">
    {media.length>0?media.map((item)=>{
      let mediaPreview; 
      if(item.media==="logo"){
      mediaPreview = <img className="external-link-img" src={item.path} height={"60"} width={"60"} alt={`${item.title}`}/>
      }else if(item.media==="image"){
      mediaPreview = <img className="external-link-img" src={item.path} height={"150"} width={"150"} alt={`${item.title}`}/>}
      else if(item.media==="video"){
        mediaPreview=<iframe width="200" height="200" src={item.path} title={item.title} 
        allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowFullScreen ></iframe>
      }
      return (
      <>
<Card border="info" style={{width:"20rem",marginTop:"5%"}} key={item.id}>
<div>
<Card.Header>
<img height={"100px"} width={"100px"} src={Logo}  alt="logo"/>
</Card.Header>
</div>
<Card.Body>
<Card.Title><strong>Title: </strong>{item.title}</Card.Title>

<ListGroup className="list-group-flush" key={item.id}>
<li className="list-group-item"><strong>Media: </strong>{item.media}</li>
<li className="list-group-item"><strong>Preview: </strong><span className="line-up">{mediaPreview}</span></li>
<li className="list-group-item"><strong>Path: </strong>{item.path}</li>
<li className="list-group-item"><strong>Id: </strong>{item.id}</li>
{item.link?<Card.Link href={item.link} className="list-group-item"><strong>Link: </strong>{item.link}</Card.Link>:null}
</ListGroup>
</Card.Body>
<Card.Footer className="action-container">
<Button variant="danger" onClick={()=>{DeleteMedia(item.id,token).then((res)=>{
  // if user wants to delete on click, media item info passed to server 
  if(res.ok===true){// if deleted 
    // alert success
    setBool(true);
    setPopUp({message:res.message,type:"Server Response",response:"text-primary"})
  }else{// if error 
    // alert error
    setBool(true);
    setPopUp({message:res.message,type:"Server Response",response:"text-danger"})
  }
})}}>Delete <Trash/></Button> 
<Button variant="warning" onClick={()=>{HandleEdit(item)}}>Edit <Folder2Open/></Button>
</Card.Footer>
</Card>
<br/>
</>
    )}):null}
{(bool===true&&newMedia===true)?<AddPageMedia token={token} bool={bool} ToggleBool={ToggleBool}/>:null}
{(bool===true&&popup.message!=="")?<PopUp message={popup.message} type={popup.type} response={popup.response} bool={bool} ToggleBool={ToggleBool}/>:null}
{(bool===true&&edit!==null)?<EditPageMedia item={edit} token={token} bool={bool} ToggleBool={ToggleBool}/>:null}

  </div>
    </>)
    
}

/*{"media":"video","path":"https://www.youtube.com/embed/kOSM9SYhAoU","title":"Hold Back The River Cover (Midnight Pacific)","id":77} */