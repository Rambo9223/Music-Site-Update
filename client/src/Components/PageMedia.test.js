/**
*@jest-environment jsdom
*/
import { render,cleanup, fireEvent, waitFor, act} from '@testing-library/react';
import * as React from 'react';
import fetchMock from 'jest-fetch-mock';
import PageMedia from './PageMedia';
import AddMediaAdmin from './Backend-Async/AddMediaAdmin';

afterEach(()=>{
    cleanup();
    jest.clearAllMocks();
});

let mockMedia = {"media":"image","path":"images/squared/Tooth&Claw_2019.jpg","title":"Tooth & Claw - 2019","id":7992}

describe('PageMedia Tests', () => {

    test('Page matches snapshot', () => {
        let page = render(<PageMedia/>);
        expect(page).toMatchSnapshot();
    });

    test('Page elements are visable', async () => {
        let page = render(<PageMedia/>);
        let addMediaButton = await page.findByText("Add Media",{exact:false});
        let pageHeader = await page.findByText("Show Media:");
        let searchButton = await page.findByText("Search",{exact:false});

        expect(addMediaButton).toBeInTheDocument();
        expect(pageHeader).toBeInTheDocument();
        expect(searchButton).toBeInTheDocument();

    });


    test('Add Media button opens modal, close button closes modal', async () => {
        let page = render(<PageMedia/>);
        let addMediaButton = await page.findByText("Add Media",{exact:false});
        expect(addMediaButton).toBeInTheDocument();
        
        await waitFor(()=>{fireEvent.click(addMediaButton);}); // open form
        let form = await page.findByTestId("modal-form");  
        let closeButton = await page.findByText("Close",{exact:false});
        expect(form).toBeInTheDocument();// confirm form 
        expect(closeButton).toBeInTheDocument();// confirm button 
        
        await waitFor(()=>{// close modal
            fireEvent.click(closeButton);
        });
        expect(form).not.toBeInTheDocument();// confirm no form
    });
        
        test('Submit form without values, expect errors and submit rejection', async () => {

            let page = render(<PageMedia/>);
            let addMediaButton = await page.findByText("Add Media",{exact:false});
            expect(addMediaButton).toBeInTheDocument();

            await waitFor(()=>{
                fireEvent.click(addMediaButton);
            }) ;
            
            let buttons = await page.findAllByRole("button");
            let submit = buttons[4];
            
            expect(submit).toBeInTheDocument();// confirm submit button 
        
            global.fetch = fetchMock.mockResponseOnce(JSON.stringify({json:mockMedia}));

            await act(()=>{return fireEvent.click(submit)});
            
            let titleError = await page.findByText("Title is required.")
            let pathError = await page.findByText("Path is required.");
    
            expect(fetch).toHaveBeenCalledTimes(0);// form is rejected by error boundries so fetch isn't called
            expect(titleError).toBeInTheDocument();// no input so error displays on form
            expect(pathError).toBeInTheDocument();

        });

        test('Submit form with values, expect submit success and mock response', async () => {

            let page = render(<PageMedia/>);
            let addMediaButton = await page.findByText("Add Media",{exact:false});
            expect(addMediaButton).toBeInTheDocument();

            await waitFor(()=>{
                fireEvent.click(addMediaButton);
            }) ;
            
            let buttons = await page.findAllByRole("button");
            let submit = buttons[4];
            expect(submit).toBeInTheDocument();// confirm submit button
        
            let select = await page.findByText("Image");// select element
            let inputs = await page.findAllByText("");
            let formInputs = inputs.filter((input)=>input.className==="form-inputs");// title & path element
            
            // pass values to input/select elements and expect values to match
            await act(async()=>{return fireEvent.input(formInputs[0],{target:{value:mockMedia.title}})});
            expect(formInputs[0].value).toBe(mockMedia.title);
    
            await act(async()=>{return fireEvent.input(formInputs[1],{target:{value:mockMedia.path}})});
            expect(formInputs[1].value).toBe(mockMedia.path);
    
            await act(async()=>{return fireEvent.select(select,{target:{value:mockMedia.media}})});
            expect(select.value).toBe(mockMedia.media);

            // arm mock response
            global.fetch = fetchMock.mockResponseOnce(JSON.stringify({json:mockMedia}));
            
            // submit form with values
            await act(()=>{return fireEvent.click(submit)});

            expect(fetch).toHaveBeenCalledTimes(1);//with values form is accepted and fetch is called 

            // mock server call with new item
            let result = await AddMediaAdmin(mockMedia,"token");
            expect(result.json).toEqual(mockMedia);

        })




    



}); 