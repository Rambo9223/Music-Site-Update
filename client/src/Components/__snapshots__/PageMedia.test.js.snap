// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageMedia Tests Page matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h3
        class="faded-title"
      >
        Page Media
      </h3>
      <div>
        <button
          class="btn btn-success"
          style="margin-top: 0.5%;"
          type="button"
        >
          Add Media 
          <svg
            fill="currentColor"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14m0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16"
            />
            <path
              d="M8 4a.5.5 0 0 1 .5.5v3h3a.5.5 0 0 1 0 1h-3v3a.5.5 0 0 1-1 0v-3h-3a.5.5 0 0 1 0-1h3v-3A.5.5 0 0 1 8 4"
            />
          </svg>
        </button>
        <h4
          class="faded-title"
        >
          Show Media:
        </h4>
        <select
          aria-label="Default select example"
          class="form-select-md"
          id="filters"
        >
          <option>
            Filter by:
          </option>
          <option
            value=""
          >
            All
          </option>
          <option
            value="image"
          >
            Images
          </option>
          <option
            value="video"
          >
            Video
          </option>
          <option
            value="logo"
          >
            Social Media
          </option>
        </select>
        <br />
        <button
          class="btn btn-primary btn btn-primary"
          style="margin-top: 0.5%;"
          type="button"
        >
          Search 
          <svg
            fill="currentColor"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14m0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16"
            />
            <path
              d="M5.255 5.786a.237.237 0 0 0 .241.247h.825c.138 0 .248-.113.266-.25.09-.656.54-1.134 1.342-1.134.686 0 1.314.343 1.314 1.168 0 .635-.374.927-.965 1.371-.673.489-1.206 1.06-1.168 1.987l.003.217a.25.25 0 0 0 .25.246h.811a.25.25 0 0 0 .25-.25v-.105c0-.718.273-.927 1.01-1.486.609-.463 1.244-.977 1.244-2.056 0-1.511-1.276-2.241-2.673-2.241-1.267 0-2.655.59-2.75 2.286m1.557 5.763c0 .533.425.927 1.01.927.609 0 1.028-.394 1.028-.927 0-.552-.42-.94-1.029-.94-.584 0-1.009.388-1.009.94"
            />
          </svg>
        </button>
      </div>
      <div
        class="gallery"
      />
    </div>
  </body>,
  "container": <div>
    <h3
      class="faded-title"
    >
      Page Media
    </h3>
    <div>
      <button
        class="btn btn-success"
        style="margin-top: 0.5%;"
        type="button"
      >
        Add Media 
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14m0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16"
          />
          <path
            d="M8 4a.5.5 0 0 1 .5.5v3h3a.5.5 0 0 1 0 1h-3v3a.5.5 0 0 1-1 0v-3h-3a.5.5 0 0 1 0-1h3v-3A.5.5 0 0 1 8 4"
          />
        </svg>
      </button>
      <h4
        class="faded-title"
      >
        Show Media:
      </h4>
      <select
        aria-label="Default select example"
        class="form-select-md"
        id="filters"
      >
        <option>
          Filter by:
        </option>
        <option
          value=""
        >
          All
        </option>
        <option
          value="image"
        >
          Images
        </option>
        <option
          value="video"
        >
          Video
        </option>
        <option
          value="logo"
        >
          Social Media
        </option>
      </select>
      <br />
      <button
        class="btn btn-primary btn btn-primary"
        style="margin-top: 0.5%;"
        type="button"
      >
        Search 
        <svg
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14m0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16"
          />
          <path
            d="M5.255 5.786a.237.237 0 0 0 .241.247h.825c.138 0 .248-.113.266-.25.09-.656.54-1.134 1.342-1.134.686 0 1.314.343 1.314 1.168 0 .635-.374.927-.965 1.371-.673.489-1.206 1.06-1.168 1.987l.003.217a.25.25 0 0 0 .25.246h.811a.25.25 0 0 0 .25-.25v-.105c0-.718.273-.927 1.01-1.486.609-.463 1.244-.977 1.244-2.056 0-1.511-1.276-2.241-2.673-2.241-1.267 0-2.655.59-2.75 2.286m1.557 5.763c0 .533.425.927 1.01.927.609 0 1.028-.394 1.028-.927 0-.552-.42-.94-1.029-.94-.584 0-1.009.388-1.009.94"
          />
        </svg>
      </button>
    </div>
    <div
      class="gallery"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
